[WIP] Implement avl add

matthiasbeyer opened the pull request:
Solves #55 .

matthiasbeyer committed:
Added implementation for avl_add() with helper

matthiasbeyer committed:
Added implementation for helper: insert_element_into_tree()

matthiasbeyer committed:
Added implementation for helper: inserting an element into the tree

matthiasbeyer committed:
Added implementation for helper: element_with_hash()

matthiasbeyer committed:
Added implementation for helper: new_avl_el()

neithernut commented:
Acceptable.
ACK.

matthiasbeyer commented:
You would leave the rotation functions unimplemented for this PR?

neithernut commented:
The interface for the rotation functions should not change. Hence, this PR can be ACK'ed without the rotation functions being merged already.
We may, however, defer the merge until the rotation functions are merged.

matthiasbeyer commented:
So, I'll add another issue for the rotation functions

TheNeikos commented on src/libreset/avl.c:
```diff
@@ -7,6 +11,170 @@ avl_alloc(void) {
+    }
+
+    if (unlikely(cmp == 0)) {
+        /* Shouldn't be possible */
```
If this is impossible, how come that it might get set to 0 earlier? Why not check directly at this point whether
`unlikely(el->hash == (*root)->hash)` is true and drop the one before?

matthiasbeyer replied:
Of course, I should simplify this algorithm here!

TheNeikos commented on src/libreset/avl.c:
```diff
@@ -7,6 +11,170 @@ avl_alloc(void) {
+        }
+    }
+
+    (*root)->height = MAX(avl_height((*root)->l), avl_height((*root)->l)) + 1;
```
Is it correct to check for Max between `x` and `x` in this case?

matthiasbeyer replied:
Nope, this was an typo, thank you for mentioning!

neithernut commented:
The rotation functions will be introduced in a seperate PR, so please remove the last commit, see PR #72. Also note that the structure of avl.c is changed by that PR.
One the rotation functions are introduced, I'd use those in the insertion and deletion functions.

matthiasbeyer committed:
Moved documentation for add()

matthiasbeyer committed:
Moved documentation for new_avl_el()

matthiasbeyer committed:
Moved documentation for insert_element_into_tree()

matthiasbeyer committed:
Moved documentation for element_with_hash()

matthiasbeyer committed:
Fix: Fixed include statements

matthiasbeyer commented:
@neithernut The PR still mentiones some issues because the rotation functions are neither defined nor implemented. Shall I wait until they are written or can I merge this right now and you'll add the implementations?

matthiasbeyer committed:
Fix: We don't need a `cmp` variable here

The case that the `cmp` variable is zero is excluded, as this would mean
that the node is already in the tree, whereas this function is assumed
NOT to be called in this case.

So we don't need to check for `cmp == 0`. Because this leaves us two
possible cases, we can simply remove the `cmp` variable and embedd the
hash comparison right into the if condition.

neithernut commented:
The PR bringing the rotation functions are on the way. See PR #74.

matthiasbeyer commented:
waiting on #74 .

neithernut commented:
Not that #74 is done, you can easily adapt your approach:
- In `insert_element_into_tree()`, you call `regen_metadata()` on the current node on your way "up" (after insertion that is). You can remove the rotations from that function. They are not needed anymore.
- In `add()` you simply call `rebalance_subtree()`, after adding a node and be done with it. The tree will be automatically rebalanced.

matthiasbeyer commented:
I will continue working on this whole stuff when I'm healthy again! Thanks for the explanation, though!

neithernut commented:
Ok. get better!

matthiasbeyer committed:
Use regen_metadata() after before ending the current recursion step. Remove rotations therefor

matthiasbeyer committed:
After adding a node to the tree, we must rebalance the appropriate subtree

neithernut commented on src/libreset/avl.c:
```diff
@@ -117,6 +178,91 @@ avl_find(
+        (*root)->r = insert_element_into_tree(el, &(*root)->r);
+    }
+
+    (*root)->height = MAX(avl_height((*root)->l), avl_height((*root)->r)) + 1;
```
regen_metadata does this for you, already.

neithernut commented on src/libreset/avl.c:
```diff
@@ -117,6 +178,89 @@ avl_find(
+        return root;
+    }
+
+    if (root->hash >= h) {
```
I'd like to see the recursion resolved the same way it's in `avl_find`, as we do not have to handle anything after recusring.
Site note: another PR may then reimplement `avl_find` by the means of this function.

matthiasbeyer committed:
Moved add() code to avl_add()

As the code is not that complex anymore, the add() function body was
moved to  the avl_add() function with this commit.

matthiasbeyer committed:
Added missing data insertion

matthiasbeyer committed:
Remove element_with_hash() helper function

The functionality of element_with_hash() is provided by avl_find(), so
use this function instead.

neithernut commented:
Nice, I completely overlooked that you can simply use `avl_find` anyway.
ACK for squash and merge.

matthiasbeyer commented:
\m/ finally! I'll do the squashing right now!
